dhall/src/Dhall/TypeCheck.hs

```diff
@@ -163,36 +156,16 @@ typeWithA tpa = loop
                     else Left (TypeError ctx e (AnnotMismatch a0 nf_A0 nf_A1))
             Nothing -> return ()
 
-        t <- loop ctx _A1
-
         let a1 = Dhall.Core.normalize a0
         let a2 = Dhall.Core.shift 1 (V x 0) a1
 
         let rest = case ls of
                 []       -> b0
                 l' : ls' -> Let (l' :| ls') b0
 
-        -- The catch-all branch directly implements the Dhall
-        -- specification as written; it is necessary to substitute in
-        -- types in order to get 'dependent let' behaviour and to
-        -- allow type synonyms (see #69). However, doing a full
-        -- substitution is slow if the value is large and used many
-        -- times. If the value being substitued in is a term (i.e.,
-        -- its type is a Type), then we can get a very significant
-        -- speed-up by doing the type-checking once at binding-time,
-        -- as opposed to doing it at every use site (see #412).
-        case Dhall.Core.normalize t of
-          Const Type -> do
```

Comment:
I removed this “fast path”, but I’d love to have a test case that behaves incorrectly when this code exists.

Reply:
Let's assume we are in a context as follows: `Vec : Natural -> Type, Vec/nil : Vec 0`. Consider the (dependently-typed) expression `let n : Natural = 0 in Vec/nil : Vec n`. The fast path is equivalent to adding `n : Natural` to the context (and forgetting about its value). But `Vec : Natural -> Type, Vec/nil : Vec 0, n : Natural |- Vec/nil : Vec n` is not valid, since the value of `n` is not known. However, the slow path, of substitution, is valid (and obviously correct): `Vec/nil : Vec 0` is the result of the substitution, and it is clearly in the context, as desired.

The big picture is that, without dependent types, the only things that influence typechecking are types (and kinds (and sorts …)), not values. Values don't matter to typechecking, only to computation. However, with dependent types, that is obviously not the case: by definition, types are now allowed to depend on values.

I think there are ways to address the concern in the comment by sharing the work more efficiently though, but it requires a different representation. The type of a term has to be associated with the node that triggered it. It starts to look like a lazy graph. I might try that approach in the PS project when I return to it over the summer. Premature optimization though.

Reply:
@MonoidMusician Those kinds of performance questions are completely solved on the [nbe-elaboration](https://github.com/dhall-lang/dhall-haskell/tree/nbe-elaboration) branch, which I hope can be merged in the summer when I get some time.